File: frontend/src/services/notifications/notification.ts
import api from "../../lib/api";

export type RecipientRuleType =
  | "everyone"
  | "grade"
  | "class"
  | "grade_section"
  | "subject"
  | "student"
  | "guardian"
  | "teacher";

export type SubjectAudience = "students" | "teachers";

export interface RecipientRule {
  type: RecipientRuleType;
  label?: string;
  grade_id?: string;
  class_id?: string;
  grade_section_id?: string;
  subject_id?: string;
  subject_audience?: SubjectAudience;
  student_id?: string;
  guardian_id?: string;
  teacher_id?: string;
}

export type NotificationCategory =
  | "general"
  | "academic"
  | "examination"
  | "attendance"
  | "timetable"
  | "events"
  | "sports"
  | "meetings"
  | "fee_reminder"
  | "emergency"
  | "discipline"
  | "holidays";

export type NotificationPriority = "normal" | "important" | "urgent";

export const CATEGORIES: { value: NotificationCategory; label: string }[] = [
  { value: "general", label: "General Announcement" },
  { value: "academic", label: "Academic" },
  { value: "examination", label: "Examination" },
  { value: "attendance", label: "Attendance" },
  { value: "timetable", label: "Timetable" },
  { value: "events", label: "Events" },
  { value: "sports", label: "Sports" },
  { value: "meetings", label: "Meetings" },
  { value: "fee_reminder", label: "Fee Reminder" },
  { value: "emergency", label: "Emergency" },
  { value: "discipline", label: "Discipline" },
  { value: "holidays", label: "Holidays" },
];

export const PRIORITIES: { value: NotificationPriority; label: string }[] = [
  { value: "normal", label: "Normal" },
  { value: "important", label: "Important" },
  { value: "urgent", label: "Urgent" },
];

export interface CreateNotificationRequest {
  title: string;
  message: string;
  category: NotificationCategory;
  priority: NotificationPriority;
  save_as_draft: boolean;
  recipient_rules: RecipientRule[];
}

export type UpdateNotificationRequest = Omit<CreateNotificationRequest, "save_as_draft">;

export interface Notification {
  id: string;
  title: string;
  message: string;
  category: NotificationCategory;
  priority: NotificationPriority;
  status: "draft" | "sent";
  recipient_rules: RecipientRule[];
  sender_name?: string;
  sent_at: string | null;
  created_at: string | null;
}

export interface NotificationStats {
  total: number;
  read: number;
  unread: number;
}

export interface MyNotification {
  recipient_id: string;
  notification_id: string;
  title: string;
  message: string;
  category: NotificationCategory;
  priority: NotificationPriority;
  sender_name: string;
  sent_at: string;
  is_read: boolean;
  is_archived: boolean;
}

export const notificationApi = {
  create: (data: CreateNotificationRequest) =>
    api.post<Notification>("/notifications", data).then((r) => r.data),

  update: (id: string, data: UpdateNotificationRequest) =>
    api.put<Notification>(`/notifications/${id}`, data).then((r) => r.data),

  send: (id: string) => api.post<Notification>(`/notifications/${id}/send`).then((r) => r.data),

  remove: (id: string) => api.delete(`/notifications/${id}`).then((r) => r.data),

  listSent: () => api.get<Notification[]>("/notifications/sent").then((r) => r.data),

  listDrafts: () => api.get<Notification[]>("/notifications/drafts").then((r) => r.data),

  stats: (id: string) => api.get<NotificationStats>(`/notifications/${id}/stats`).then((r) => r.data),

  listMine: () => api.get<MyNotification[]>("/me/notifications").then((r) => r.data),

  listMyArchived: () => api.get<MyNotification[]>("/me/notifications/archived").then((r) => r.data),

  unreadCount: () =>
    api.get<{ unread_count: number }>("/me/notifications/unread-count").then((r) => r.data.unread_count),

  markRead: (notificationId: string) =>
    api.post(`/me/notifications/${notificationId}/read`).then((r) => r.data),

  archive: (notificationId: string) =>
    api.post(`/me/notifications/${notificationId}/archive`).then((r) => r.data),

  unarchive: (notificationId: string) =>
    api.post(`/me/notifications/${notificationId}/unarchive`).then((r) => r.data),
};

File: backend/internal/models/notifications/notification.go
package notifications

import (
	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgtype"
)

type RecipientRuleType string

const (
	RuleEveryone     RecipientRuleType = "everyone"
	RuleGrade        RecipientRuleType = "grade"
	RuleClass        RecipientRuleType = "class"
	RuleGradeSection RecipientRuleType = "grade_section"
	RuleSubject      RecipientRuleType = "subject"
	RuleStudent      RecipientRuleType = "student"
	RuleGuardian     RecipientRuleType = "guardian"
	RuleTeacher      RecipientRuleType = "teacher"
)

type SubjectAudience string

const (
	AudienceStudents SubjectAudience = "students"
	AudienceTeachers SubjectAudience = "teachers"
)

// RecipientRule is one targeting criterion picked in the composer; a
// notification combines one or more of these, unioned together at send
// time. Label is display-only (e.g. "Grade 10"), filled in by the
// frontend so a sent notification's audience can be shown back without
// re-resolving IDs to names.
type RecipientRule struct {
	Type            RecipientRuleType `json:"type"`
	Label           string            `json:"label,omitempty"`
	GradeID         *uuid.UUID        `json:"grade_id,omitempty"`
	ClassID         *uuid.UUID        `json:"class_id,omitempty"`
	GradeSectionID  *uuid.UUID        `json:"grade_section_id,omitempty"`
	SubjectID       *uuid.UUID        `json:"subject_id,omitempty"`
	SubjectAudience SubjectAudience   `json:"subject_audience,omitempty"`
	StudentID       *uuid.UUID        `json:"student_id,omitempty"`
	GuardianID      *uuid.UUID        `json:"guardian_id,omitempty"`
	TeacherID       *uuid.UUID        `json:"teacher_id,omitempty"`
}

type CreateNotificationRequest struct {
	Title          string          `json:"title" binding:"required"`
	Message        string          `json:"message" binding:"required"`
	Category       string          `json:"category" binding:"required"`
	Priority       string          `json:"priority" binding:"required"`
	SaveAsDraft    bool            `json:"save_as_draft"`
	RecipientRules []RecipientRule `json:"recipient_rules" binding:"required,min=1"`
}

type UpdateNotificationRequest struct {
	Title          string          `json:"title" binding:"required"`
	Message        string          `json:"message" binding:"required"`
	Category       string          `json:"category" binding:"required"`
	Priority       string          `json:"priority" binding:"required"`
	RecipientRules []RecipientRule `json:"recipient_rules" binding:"required,min=1"`
}

type NotificationResponse struct {
	ID             uuid.UUID          `json:"id"`
	Title          string             `json:"title"`
	Message        string             `json:"message"`
	Category       string             `json:"category"`
	Priority       string             `json:"priority"`
	Status         string             `json:"status"`
	RecipientRules []RecipientRule    `json:"recipient_rules"`
	SenderName     string             `json:"sender_name,omitempty"`
	SentAt         pgtype.Timestamptz `json:"sent_at"`
	CreatedAt      pgtype.Timestamptz `json:"created_at"`
}

type NotificationStatsResponse struct {
	Total  int32 `json:"total"`
	Read   int32 `json:"read"`
	Unread int32 `json:"unread"`
}

type MyNotificationResponse struct {
	RecipientID    uuid.UUID          `json:"recipient_id"`
	NotificationID uuid.UUID          `json:"notification_id"`
	Title          string             `json:"title"`
	Message        string             `json:"message"`
	Category       string             `json:"category"`
	Priority       string             `json:"priority"`
	SenderName     string             `json:"sender_name"`
	SentAt         pgtype.Timestamptz `json:"sent_at"`
	IsRead         bool               `json:"is_read"`
	IsArchived     bool               `json:"is_archived"`
}

var ValidCategories = map[string]bool{
	"general": true, "academic": true, "examination": true, "attendance": true,
	"timetable": true, "events": true, "sports": true, "meetings": true,
	"fee_reminder": true, "emergency": true, "discipline": true, "holidays": true,
}

var ValidPriorities = map[string]bool{"normal": true, "important": true, "urgent": true}
